<a href="https://colab.research.google.com/github/iskyd/music-intonation/blob/master/Intonation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Gli intervalli più semplici sono l'unisono, con un ratio di 1:1, e l'ottava, con un ratio di 2:1
Questo vuol dire che una corda, nelle stesse condizioni, lunga 1/2 di un'altra corda produrrà un suono un'ottava più alta (allo stesso modo una corda lunga il doppio produrrà un suono un'ottava più bassa).




In [0]:
root = 220
octave = 220 * 2

print('Root', root)
print('Octave', octave)

Root 220
Octave 440


Il prossimo ratio più semplice è il ratio 3:2 che rappresenta la quinta. Allo stesso modo dell'ottava quindi una corda lunga 2/3 di un'altra produrrà un suono un'ottava più alta.



In [0]:
root = 220
fifth = 220 * 3/2

print('Root', root)
print('Fifth', fifth)

Root 220
Fifth 330.0


La scala Pitagorica nasce proprio da questo ratio, 3/2, da cui nascono tutte le altre note. L'intervallo di quinta perfetta è il più forte intervallo consonante e per questo è stato usato come partenza.

Tutto era basato sulla tetractys, ovvero sui primi 4 numeri naturale (1,2,3,4), infatti i ratio da cui la scala deriva sono 1:1 (unisono), 2:1 (ottava), 3:2 (quinta), 4:3 (quarta).

La scala veniva generata quindi con una successione di quinte perfette (circolo delle quinte).



In [0]:
root = 220
notes = [root]

def reduce(note):
  while(note > root * 2):
    note = note / 2
   
  return note

note = root
for i in range(1, 11):
  note = note * (3/2)
  notes.append(reduce(note))  

notes.append(440)
notes.sort()
print(notes)



[220, 234.931640625, 247.5, 264.298095703125, 278.4375, 313.2421875, 330.0, 352.3974609375, 371.25, 396.4471435546875, 417.65625, 440]


Ogni 12 intervalli di quinta torniamo sulla stessa fondamentale 7 ottave più in alto.
La successione di questi intervalli di quinta però non coincide perfettamente con l'ottava.
La differenza tra le due frequenze è chiamato comma.

In [0]:
root = 100

fifth = root * (3/2)**12
octave = root * 2 ** 7

comma = fifth - octave

print('Comma', comma)

Comma 174.6337890625


L'intonazione naturale (Just intonation) nasce dalla successione naturale dei suoni armonici.
Quando due intervalli ad intonazione naturale vengono suonati non si crea alcuna interferenza tra loro e si combinano perfettamente.
A differenza della scala pitagiroca vengono anche definiti i ratio della terza maggiore, 5/4, e della terza minore, 6/5. Tutte le frazioni (3/2, 4/3, 5/4, 6/5) sono numeri superpartientes, il numeratore supera di 1 il denominatore.
Gli altri intervalli derivano dalla differenza con questi intervalli principali.
Gli intervalli di terza e di sesta risultano essere perfettamente consonanti a differenza degli stessi intervalli nel temperamento pitagorico.


*   Unisono: 1:1
*   Seconda maggiore: 9/8
*   Terza maggiore: 5/4
*   Quarta giusta: 4/3
*   Quinta giusta: 3/2
*   Sesta maggiore: 5/3
*   Settima maggiore: 15/8
*   Ottava: 2:1



In [0]:
ratios = [
    1/1,
    9/8,
    5/4,
    4/3,
    3/2,
    5/3,
    15/8,
    2/1
]

root = 220

scale = [root * ratio for ratio in ratios]

print(scale)
  

[220.0, 247.5, 275.0, 293.3333333333333, 330.0, 366.6666666666667, 412.5, 440.0]


La distanza tra due intervalli si esprime in cent. 
Il cent è la 1200esima parte di un'ottava. Questa misura nasce dal sistema temperato equabile in ci sono 12 semitono oguno alla stessa distanza dall'altro, distanza di 100 cent (1200 / 12).
La distanza in cent si trova applicando la seguente formula:

In [0]:
import math

f1 = 220
f2 = 440

cent = 1200 * math.log(f2 / f1, 2) # f2 > f1
print(cent, 'cent')

Distanza tra gli intervalli puri e gli intervalli generati dalla scala pitagorica.
Gli intervalli di terza maggiore, quarta giusta, sesta maggiore e settima maggiore risultano più ampi nella scala pitagorica.


In [0]:
import math

pure_intervals = [220.0, 247.5, 275.0, 293.3333333333333, 330.0, 366.6666666666667, 412.5, 440.0]
pythagorean_intervals = [220, 247.5, 278.4375, 297.3353576660156, 330.0, 371.25, 417.65625, 440]

for key, pure_frequency in enumerate(pure_intervals):
  print('pure {} - pythagorean {} = {} cents'.format(pure_frequency, pythagorean_intervals[key], 1200 * math.log(pythagorean_intervals[key] / pure_frequency, 2)))

pure 220.0 - pythagorean 220 = 0.0 cents
pure 247.5 - pythagorean 247.5 = 0.0 cents
pure 275.0 - pythagorean 278.4375 = 21.50628959671478 cents
pure 293.3333333333333 - pythagorean 297.3353576660156 = 23.460010384649014 cents
pure 330.0 - pythagorean 330.0 = 0.0 cents
pure 366.6666666666667 - pythagorean 371.25 = 21.50628959671478 cents
pure 412.5 - pythagorean 417.65625 = 21.50628959671478 cents
pure 440.0 - pythagorean 440 = 0.0 cents


Prendiamo ora gli intervalli a distanza di tono della scala pura (seconda maggiore) e vediamo la distanza che c'è tra loro.

In [0]:
import math

print('root - 2nd: {} = {} cents'.format(247.5 / 220, 1200 * math.log(247.5 / 220, 2)))
print('2nd - 3rd: {} = {} cents'.format(275 / 247.5, 1200 * math.log(275 / 247.5, 2)))
print('4th - 5th: {} = {} cents'.format(330 / 293.3333333333333, 1200 * math.log(330 / 293.3333333333333, 2)))
print('5th - 6th: {} = {} cents'.format(366.6666666666667 / 330, 1200 * math.log(366.6666666666667 / 330, 2)))
print('6th - 7th: {} = {} cents'.format(412.5 / 366.6666666666667, 1200 * math.log(412.5 / 366.6666666666667, 2)))

root - 2nd: 1.125 = 203.91000173077484 cents
2nd - 3rd: 1.1111111111111112 = 182.40371213406007 cents
4th - 5th: 1.125 = 203.91000173077484 cents
5th - 6th: 1.1111111111111112 = 182.40371213406007 cents
6th - 7th: 1.125 = 203.91000173077484 cents


Come possiamo vedere abbiamo due diversi tipi di tono: uno di 9/8 (1.125, 203,9 cents) e uno di 10/9 (1.1111111111111112, 182,4 cents)

Per questo motivo l'utilizzo della scala pura è limitato a una sola tonalità.
Prendiamo come root il quinto grado ovvero 330hz e calcoliamo il suo secondo grado

In [0]:
print('6th of A', 366.6666666666667)
print('2nd of E', 330 * (9/8))

6th of A 366.6666666666667
2nd of E 371.25


Come vediamo il F# (sesto maggiore di A e secondo maggiore di E) ha una frequenza differente in base alla tonalità di partenza. Infatti come sesto grado maggiore di A ha una frequenza di 366.6666666666667 mentre come secondo maggiore di E ha una frequenza di 371.25. Questi sono infatti rispettivamente tono minore e tono maggiore.

Questo limita l'utilizzo della scala pura ha una sola tonalità. Un grosso problema per gli strumenti ad intonazione fissa come ad esempio il pianoforte, accordarsi ogni cambio di tonalità non è infatti pensabile.
Discorso diverso invece per gli strumenti senza un'intonazione fissa come ad esempio il violino o gli strumenti a fiato.

Tornando ai due tipi di tono differente all'interno della scala pura possiamo quindi affermare che esistono:


*   Tono maggiore: 9/8
*   Tono minore: 10/9

Proviamo a vedere ora il rapporto per quanto riguarda il semitono, che troviamo tra il terzo grado maggiore e il quarto grado giusto e tra il settimo grado maggiore e l'ottava.



In [0]:
print('3rd - 4th: {} = {} cents'.format(293.3333333333333/275, 1200 * math.log(293.3333333333333/275, 2)))
print('7th - 8th: {} = {} cents'.format(440/412.5, 1200 * math.log(440/412.5, 2)))

3rd - 4th: 1.0666666666666667 = 111.73128526977776 cents
7th - 8th: 1.0666666666666667 = 111.73128526977776 cents


La distanza di semitono è la stessa.

Vediamo ora il rapporto tra due note a distana di tono, come fatto per la scala pura, della scala pitagorica.

In [0]:
print('root - 2nd: {} = {} cents'.format(247.5 / 220, 1200 * math.log(247.5 / 220, 2)))
print('2nd - 3rd: {} = {} cents'.format(278.4375 / 247.5, 1200 * math.log(278.4375 / 247.5, 2)))
print('4th - 5th: {} = {} cents'.format(330.0 / 297.3353576660156, 1200 * math.log(330.0 / 297.3353576660156, 2)))
print('5th - 6th: {} = {} cents'.format(371.25 / 330, 1200 * math.log(371.25 / 330, 2)))
print('6th - 7th: {} = {} cents'.format(417.65625 / 371.25, 1200 * math.log(417.65625 / 371.25, 2)))

root - 2nd: 1.125 = 203.91000173077484 cents
2nd - 3rd: 1.125 = 203.91000173077484 cents
4th - 5th: 1.1098579146132872 = 180.44999134612573 cents
5th - 6th: 1.125 = 203.91000173077484 cents
6th - 7th: 1.125 = 203.91000173077484 cents


Possiamo vedere che tutti gli intervalli di tono hanno lo stesso rapporto a differenza di quello tra quarta e quinta che risulta essere più piccolo.
Facciamo la stessa cosa per gli intervalli di semitono

In [0]:
print('3rd - 4th: {} = {} cents'.format(297.3353576660156 / 278.4375, 1200 * math.log(297.3353576660156 / 278.4375, 2)))
print('7th - 8th: {} = {} cents'.format(440 / 417.65625, 1200 * math.log(440 / 417.65625, 2)))

3rd - 4th: 1.06787109375 = 113.68500605771193 cents
7th - 8th: 1.0534979423868314 = 90.22499567306306


In questo caso anche gli intervalli di semitono risultano esseree differenti di circa 23,4 cents.


Nella just intonation il valore in cent dei possibili intervalli è

In [0]:
import math

print('Tono maggiore:', 1200 * math.log(247.5 / 220, 2))
print('Tono minore:', 1200 * math.log(275/247.5, 2))
print('Semitono:', 1200 *  math.log(293.3333333333333/275, 2))

Tono maggiore 203.91000173077484
Tono minore 182.40371213406007
Semitono 111.73128526977776


La just intonation, oltre a non poter essere utilizzata con la stessa intonatura in più tonalità, non permette di ottenere tutti gli accordi con intervalli puri neanche all'interno della stessa tonalità.
Prendiamo l'intervallo tra D (247.5hz) e F (293.3333333333333hz), terza minore, che come abbiamo visto deve essere di 6/5, per essere pura.

In [0]:
print('Cents D - F:', 1200 *  math.log(293.3333333333333/247.5, 2))
print('Cents pure interval:', 1200 *  math.log(6/5, 2))

Cents D - F 294.13499740383764
Cents pure interval 315.64128700055255


Il meantone temperament è un sistema di intonazione in cui vengono modificate le quinte per migliorare le terze. Il meantone temperament più famoso è il quarter-comma meantone, in cui le quinte vengono modificate di 1/4 del valore del syntonic comma che corrisponde alla differenza tra l'intervallo di terza puro, 5/4, e quello ottenuto nella scala pitagorica (pythagorean tuning), 81:64.

In [0]:
import math

pythagorean_third = ((3/2)**4) / 4 # Quattro quinte ascendenti formano un intervallo di terza A - E - B - F# - C# -> A - C# 
pure_third = 5/4

print('Syntonic comma:', pythagorean_third / pure_third) # 81:80
print('Syntonic comma cents:', 1200 * math.log(pythagorean_third / pure_third, 2))

Syntonic comma: 1.0125
Syntonic comma cents: 21.50628959671478


Quattro intervalli di quinta consecutivi formano un intervallo di terza: A - E - B - F# - C# -> A - C#
Come abbiamo visto nel pythagorean tuning l'intervallo di terza non è puro, ma modificando le quinte ognuna di 1/4 di syntonic comma otteniamo un intervallo A - C# puro.

In [0]:
syntonic_comma = 81/80


root = 220

pure_third = 5/4 * root

i = 0

def reduce(note):
  while(note > root * 2):
    note = note / 2
   
  return note
    
note = root

while(i < 4):
  note = note * ((3/2) / (syntonic_comma ** (1/4)))
  i += 1

print('Pure third:', pure_third)
print('quarter-comma meantone third:', reduce(note))

Pure third: 275.0
quarter-comma meantone third: 275.0


Calcoliamo ora le 12 note partendo dalla frequenza 220

In [0]:
root = 220
syntonic_comma = 81/80

def reduce(note):
  while(note > root * 2):
    note = note / 2
    
  return note

notes = [root]
note = root

for i in range(1, 11):
  if i < 5:
    note = note * ((3/2) / (syntonic_comma ** (1/4)))
  else:
    note = note * (3/2)
    
  notes.append(reduce(note))

notes.append(440) #octave

notes.sort()
print(notes)

[220, 232.03125, 245.96747752497689, 261.03515625, 275.0, 309.375, 328.97673186866854, 348.046875, 367.80716773703216, 391.552734375, 412.5, 440]
